### PR TITLE
feat(marks): marks keep "view" info and restore it on jump

### PR DESCRIPTION
**Pull request by muniter**
# Summary

## Refactor

Previously most functions used to "get" a mark returned a position,
changed the line number and sometimes changed even the current buffer.

Now functions return a {x}fmark_T making calling context aware whether
the mark is in another buffer without arcane casting. A new function is
provided for switching to the mark buffer and returning a flag style
Enum to convey what happen in the movement. If the cursor changed, line,
columns, if it changed buffer, etc.

The function to get named mark was split into multiple functions.

- `mark_get() -> fmark_T`
- `mark_get_global() -> xfmark_T`
- `mark_get_local() -> fmark_T`
  - `mark_get_motion() -> fmark_T`
  - `mark_get_visual() -> fmark_T`

Functions that manage the changelist and jumplist were also modified to
return mark types.

- `get_jumplist -> fmark_T`
- `get_changelist -> fmark_T`

The refactor is also seen mainly on normal.c, where all the mark
movement has been siphoned through one function nv_gomark, while the
other functions handle getting the mark and setting their movement
flags. To handle whether context marks should be left, etc.

## Mark View

While doing the refactor the concept of a mark view was also
implemented:

The view of a mark currently implemented as the number of lines between
the mark position on creation and the window topline. This allows for
moving not only back to the position of a mark but having the window
look similar to when the mark was defined. This is done by carrying and
extra element in the fmark_T struct, which can be extended later to also
restore horizontal shift.

### User space features

1. There's a new option, jumpoptions+=view enables the mark view restoring
automatically when using the jumplist, changelist, alternate-file and
mark motions. ```<C-O> <C-I> g; g, <C-^> '[mark] `[mark]```

## Limitations

- The view information is not saved in shada.
- Calls to `get_mark()` should copy the value in the pointer since we are
  using `pos_to_mark()` to wrap and provide a homogeneous interfaces. **This
  was also a limitation in the previous state of things**.

## Old content

<details>
<summary>Previous <b>long comments</b></summary>

I'm setting this PR as ready for review, what's in mark.c is an interface I like that provides a nice abstraction and makes marks feel more homogeneous, and an easier path to extensibility. Same for the changes in other files.

What has been done:

- [x] **Refactoring:**
  - [x] Refactor getting marks `mark_get`. To only get marks, unlike the previous version that was getting, moving, switching buffer, etc. The implementation logic wise is very similar to the previous version, but works with marks instead of just positions, to support further extensibility.
    - [x] Provide multiple functions for the different kinds of marks `mark_get_local, mark_get_global, mark_get_motion, mark_get_visual` (explanation in docstring).
    - [x] Provide a way to handle positions as marks `pos_to_mark` to make code easier to follow and with less paths.
    - [x] Provide a way of getting a mark and not resolving the fnum, meaning not loading the buffer it ***might*** be on, this is useful for API functions that want to provide the mark information but not load the buffer for performance.
    - [x] De*curbuf*y and de*curwin*y.
  - [x] New function `mark_move_to` to move to the given mark, providing various functionality (leave a context mark, leave a context mark only if switching buffer, move to beginning of line, restore the mark "view", etc). Returns in flags style to convey all the information the caller side would want, allowing not having to track things like the cursor, buffer before, etc; which sometimes was done with arcane casting (`(pos_T *)-1` is your friend). The flexibility of this function has also allowed to consolidate the "moving to a mark" in most of the codebase to this one function (didn't touch `tags.c` logic).
  - [x] Refactor multiple functions on `mark.c` to operate on marks instead of positions, which allows consolidating code paths and extensibility. `changelist, jumplist, mark_check, getnextmark`. And also remove the part were **they were switching the buffer and moving the cursor**, to instead use `mark_move_to`.
  - [x] Make `mark_check` to check marks not positions and be the only place errors are set (apart from jumplist and changelist which have different handling).
  - [x] Update named marks API functions to take advantage of the improved behavior.
  - [x] Refactor `normal.c` to use the new mark functions which allows cleaning things up, less code paths by wrapping `mark_move_to` with `nv_mark_move_to` to also handle the operator logic stuff.
  - [x] Fix `change.c` not setting the last change mark fnum (buffer number) (important in the new codepath).
  - [x] Do not break compatibility (AFAIK and tested it's all ok, otherwise is a regression).
  - [x] Tests since I needed granularity to be sure things were working as expected, also found the oldtest lacking not catching some functional things.
- [x] **Feature:** save the "view" the mark was defined on and provide ways to restore it. The feature is currently simple, save the offset between the mark lnum and the topline at the moment of definition to use when restoring the view. Which is naive with all kind of virtual lines, but works pretty well IMHO, fallback to previous behavior when view can't be restored.
  - [x] Extend the structure of `fmark_T` and therefore all the setting/resetting macros to accommodate the view information.
  - [x] Define `fmarkv_T` to contain the view information, currently only `topline_offset` field. Went for a struct thinking somethings might be needed to handle virtual/filler lines.
  - [x] Provide functions for the feature (`mark_view_make, mark_view_restore`)
  - [x] New option option `jumpoptions=view` which will restore the mark view when navigating the jumplist and changelist.
    - [x] On changelist only save the view if the change is visible in the window
  - [x] Provide a way to restore view of named marks. I don't think doing this with an options is a good idea since 'a,\`a are motions and caused the view moving with some plugins like **unimpaired** with `[<space>, ]<space>` so I **propose two alternatives, whatever the maintainers prefer or none at all**.
    - [x] Provide z\` and z' to move to the previous used named mark but also set the view, this follows conventions of "z" family setting the view. So one can do `'az'` which rolls easy on the fingers.
    - [x] Provide `[count]'` for moving to a mark and also setting the view, this also seems very idiomatic whatever people think is best I'm happy with any choice. This has the extra advantage of `nnoremap ' 1'` for people that want the view behavior always.
  - [x] Test for the new feature.
    - [ ] Add even more tests (changelist)
  - [ ] Documentation (once the refactor and approach is considered good).
  - [ ] SHADA: save the new data (perhaps could be done in a followup?).

**Also would like to point out that not much code has been added, I'd say between test and function documentation there's about 700 lines which would be around net 0~100.**

Why a new feature: I find disorienting jumping around especially for quick jumps when you know where things are but the jump esentially changes were things are by scrolling the view. There's also no easy way to capture the way something looks (cursor position and view) and restore it.

What's needed:

- [ ] Comments from maintainers if they like the feature and the approach.

What this PR might help in the future:

- Opening files on new sessions in the same position and view they were left on, useful when:
  - Buffer switching (keep a mark per buffer per window to restore, makes switching buffer like switching windows experience).
  - Sessions restoring.
- Change the named mark tracking engine to extmarks!.
- Introduce arbitrary named marks, having bookmarks!? (pretty straightforward now)
- Change the just merged named marks API functions to support the view definition/reading, before it's released in 0.6
- Clean mark.c a bit more, I don't see the need of `xfmark_T` is just two fields that can leave in `fmark_T` and simplifies the handling of marks, a mark with fnum is not resolved and should have a fname.

Things to change:

The logic to restore view for the changelist needs more work, since changes can be done out of view that create new marks with incorrect views.

<details>
  <summary>Initial comments</summary>

This PR attemtpts:

1. Untangle mark.c which has the following issues:
    - Only one function to get all type of marks, with multiple side effects and returns only a position `pos_T`.
    - Changes the cursor: Instead of changing it return the mark as the name implies and let `normal.c` function handle the moving.
    - Changes the current buffer **sometimes**: Instead return global/file  marks `xfmark_T` and provide a clear function that changes the buffer and return a `bool` wether it was changed or not.
    - Returns only a position `pos_T` making it harder to extend the concept of marks: if the function is called `getmark` it should actually get the mark not a position. Now provide different functions for every type of mark that return mark structs and provide a helper to know which function to call.
    - Has "motion style marks" mixed in with actual marks: Some marks like `{, }, <, >, (, )` are not marks, some are just positions, others are motions, abstract this away on a function.
    - All this complicated way of working leads to awkard extensibility like what can be seen in `nvim_buf_get_mark` [having to switch the buffer](https://github.com/muniter/neovim/blob/7715f47d48701c05a63c25804809d11cef71cda1/src/nvim/api/buffer.c#L1148) and then restore to try and get a mark position.

1. Add support for saving a **"view"** in the mark definition, to be able to return to the position and the view it was on.
    - This is done by creating a new field in `fmarkv_T view;` (bad name, suggestions welcome 😄 ) which is a struct with view information, currently topline, and botline. I decided to use a struct instead of just adding field since more fileds are neccessary for a followup supporting more correct behavior with `plines` aka ghostly text and the complex horizontal scrolling aka `boguscols`.

1. Update mark setter functions/macros to support adding view information
1. Refactor `normal.c` usage of marks to reflect the new behavior, and also concentrate/reduce locations where setting the cursor from a mark, allowing to implement the logic of restoring the view there.
1. Provide an option where to hide the restoring of view when using marks, to be **opt-in**.
1. Modify shada to also save this new data 🤨
1. **A good amount of tests since this PR is touching quite a few functions that are used a lot in editor features.**

Will keep updating this main post with a todo list soon, and the things I surely forgot to mention. Also if you know of a better name for this PR please change it 😄 

The PR is still rough around the edges, with some todo's and comments that have to be changed as well.

</details>

</details>

Here's a quick demo:

https://user-images.githubusercontent.com/9699804/135173357-ed7bf875-d783-4ded-81fe-b16b9f86b0bf.mp4

### CODE REVIEW

File: src/nvim/mark.c [hunk 1]
```diff
@@ -67,6 +69,7 @@ int setmark(int c)
 /// Free fmark_T item
 void free_fmark(fmark_T fm)
 {
+
```
Comment (gpanders) [inline src/nvim/mark.c:72]:
```suggestion
```

File: src/nvim/mark.c [hunk 2]
```diff
@@ -443,24 +451,238 @@ pos_T *getmark_buf_fnum(buf_T *buf, int c, bool changefile, int *fnum)
   return posp;
 }
 
+/// Get they type of a given mark.
```
Comment (gpanders) [inline src/nvim/mark.c:454]:
```suggestion
/// Get the type of a given mark.
```

File: src/nvim/normal.c [hunk 3]
```diff
@@ -6443,28 +6444,16 @@ static void n_swapchar(cmdarg_T *cap)
   }
 }
 
-/*
- * Move cursor to mark.
- */
-static void nv_cursormark(cmdarg_T *cap, int flag, pos_T *pos)
+/// Move the cursror to the mark position
```
Comment (gpanders) [inline src/nvim/normal.c:6447]:
```suggestion
/// Move the cursor to the mark position
```

**Review by gpanders — CHANGES_REQUESTED** [review]
This looks really good to me so far. Can you update the option description of `'markview'`? It's not clear at all to me from the description what it does. You'll also need to add it to `options.txt`.

**Comment by muniter** [comment]
@gpanders Thanks for taking a look, currently still on the process of refactoring, and adding tests to make sure I'm not breaking something! It might be a good idea to just limit this PR to the refactor, and then introduce the new feature in another, since it's so much, will see.

**Comment by muniter** [comment]
@gpanders test are passing for the hard part  🚀🎉😁

**Comment by muniter** [comment]
CI failures are related to the node provider, maybe network problems? Test are passing on my fork.

**Comment by chentau** [comment]
This will make the implementation of #16067 a lot simpler: the problem was that `getmark_buf` only returns a `pos_T` and not the full `fmark_T` structure, which this PR resolves. Overall seems like a very useful refactor.

**Comment by muniter** [comment]
Rebased, went over all C changes again, polished some things, reworded some names and comments.

**Comment by gpanders** [comment]
> Provide z` and z' to move to the previous used named mark but also set the view, this follows conventions of "z" family setting the view. So one can do 'az' which rolls easy on the fingers.
>Provide [count]' for moving to a mark and also setting the view, this also seems very idiomatic whatever people think is best I'm happy with any choice. This has the extra advantage of nnoremap ' 1' for people that want the view behavior always.

What about a combination of the two? Make `z'{mark}` and ``z`{mark}`` jump to `{mark}` and also reset the view. This way users can `nnoremap ' z'` (for example).

File: src/nvim/option_defs.h [hunk 4]
```diff
@@ -492,9 +492,10 @@ EXTERN int p_js;                // 'joinspaces'
 EXTERN char_u *p_jop;           // 'jumpooptions'
 EXTERN unsigned jop_flags;
 #ifdef IN_OPTION_C
-static char *(p_jop_values[]) = { "stack", NULL };
+static char *(p_jop_values[]) = { "stack", "view", NULL };
```
Comment (gpanders) [inline src/nvim/option_defs.h:495]: Need to update the help text in the docs for jumpoptions

File: test/functional/editor/jump_spec.lua [hunk 5]
```diff
@@ -137,3 +139,88 @@ describe("jumpoptions=stack behaves like 'tagstack'", function()
...
+    14 line     |
+    15 line     |
+    16 line     |
+    17 line     |
+    18 line     |
+                |
+    ]])
+  end)
+
+  it('restores  the view across files', function()
```
Comment (gpanders) [inline test/functional/editor/jump_spec.lua:182]:
```suggestion
  it('restores the view across files', function()
```

File: test/functional/editor/jump_spec.lua [hunk 6]
```diff
@@ -137,3 +139,88 @@ describe("jumpoptions=stack behaves like 'tagstack'", function()
...
+    screen:expect([[
+    ^12 line     |
+    13 line     |
+    14 line     |
+    15 line     |
+                |
+    ]])
+  end)
+
+  it('fallback to standard behavior when view can\'t be recovered', function()
```
Comment (gpanders) [inline test/functional/editor/jump_spec.lua:206]:
```suggestion
  it('falls back to standard behavior when view can\'t be recovered', function()
```

File: test/functional/editor/jump_spec.lua [hunk 7]
```diff
@@ -137,3 +139,88 @@ describe("jumpoptions=stack behaves like 'tagstack'", function()
...
+    local screen = Screen.new(5, 8)
+    screen:attach()
+    command("edit " .. file1)
+    feed("7GzbG")
+    curbufmeths.set_lines(0, 2, true, {})
+    -- Since lines where deleted can't se a 7 lines topline offset, so it
+    -- fallback to centering the view.
```
Comment (gpanders) [inline test/functional/editor/jump_spec.lua:213]: Can you clarify this comment? I don't follow, specifically "can't se a 7 lines topline offset"

File: src/nvim/option_defs.h (hunk 4 above)
Comment (muniter) [reply src/nvim/option_defs.h:495]: Let me know what you think https://github.com/muniter/neovim/blob/4897a954bc06b3d6935b53b783285f5994e48f45/runtime/doc/options.txt#L3513

File: test/functional/editor/jump_spec.lua (hunk 7 above)
Comment (muniter) [reply test/functional/editor/jump_spec.lua]: I think this explanation it's a little better https://github.com/muniter/neovim/blob/4897a954bc06b3d6935b53b783285f5994e48f45/test/functional/editor/jump_spec.lua#L212

File: runtime/doc/options.txt [hunk 8]
```diff
@@ -3510,6 +3510,12 @@ A jump table for the options with a short description can be found at |Q_op|.
 			jumplist and then jumping to a location.
 			|jumplist-stack|
 
+	  view          When moving on the jumplist try to recover the view in
+			which the jump was defined. The view of a jump is
```
Comment (gpanders) [inline runtime/doc/options.txt:3514]:
```suggestion
	  view          When moving through the jumplist try to restore the view from
			when the jump occurred. The view of a jump is
```

**Comment by gpanders** [comment]
@muniter Once the `z'{mark}` change is done this LGTM. I'll leave it a few days for someone else to review (@bfredl? @zeertzjq?).

**Comment by muniter** [comment]
Perhaps @jamessan which I've seen work on some named mark related things :)

File: test/functional/editor/mark_spec.lua [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+                  |
+      ]])
+  end)
+
+  it('is restored with [count]\' and [count]`', function()
+      local screen = Screen.new(5, 8)
+      screen:attach()
+      command("edit " .. file1)
+      feed("<C-e>jWma")
+      feed("G1'a") -- move away, come back and use z' to restore the view
```
Comment (jamessan) [inline test/functional/editor/mark_spec.lua:339]: This isn't using `z'`

File: test/functional/editor/mark_spec.lua [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+      ^3 line      |
+      4 line      |
+      5 line      |
+      6 line      |
+      7 line      |
+      8 line      |
+                  |
+      ]]
+      screen:expect({grid=expected})
+      feed("G1`a") -- move away and use z` to comeback and restore the view
```
Comment (jamessan) [inline test/functional/editor/mark_spec.lua:351]: Ditto

**Review by jamessan — COMMENTED** [review]
Where's the documentation about the new user commands (`z'`, etc.)?

**Comment by muniter** [comment]
@jamessan thanks for the review, there's definitely some foot guns to fix 😁 I'll address the comments as soon as I have time again.

File: src/nvim/mark_defs.h [hunk 11]
```diff
@@ -10,6 +10,34 @@
  * (a normal mark is a lnum/col pair, the same as a file position)
  */
 
+/// Flags for outcomes when moving to a mark.
+typedef enum {
+  kMarkMoveSuccess = 1,  /// Successful move.
```
[the remark this answers is not in the document]
Comment (muniter) [reply src/nvim/mark_defs.h:15]: I went by the [dev_style](https://github.com/muniter/neovim/blob/825727875b081a9ace8dce251b9536bee7eebe3d/runtime/doc/dev_style.txt#L551) examples.

File: src/nvim/mark_defs.h [hunk 12]
```diff
@@ -25,24 +53,39 @@
...
+/// Represetns view in which the mark was created
+typedef struct fmarkv {
+  linenr_T topline_offset;  // Amount of lines from the mark lnum to the top of the window.
+} fmarkv_T;
+
+#define INIT_FMARKV { 0 }
+
 /// Structure defining single local mark
 typedef struct filemark {
-  pos_T mark;           ///< Cursor position.
```
[the remark this answers is not in the document]
Comment (muniter) [reply src/nvim/mark_defs.h:36]: I went by the [dev_style](https://github.com/muniter/neovim/blob/825727875b081a9ace8dce251b9536bee7eebe3d/runtime/doc/dev_style.txt#L551) examples. Since I added something I went for the same style, I'm happy to revert.

File: test/functional/editor/mark_spec.lua (hunk 9 above)
Comment (muniter) [reply test/functional/editor/mark_spec.lua:339]: 👍🏽

File: test/functional/editor/mark_spec.lua (hunk 10 above)
Comment (muniter) [reply test/functional/editor/mark_spec.lua:351]: 👍🏽

File: src/nvim/mark.c [hunk 13]
```diff
@@ -221,245 +228,402 @@ void checkpcmark(void)
...
+
+/// Wrap a pos_T into an fmark_T, used to abstract marks hadnling.
+/// @note  view fields are set to 0.
+/// @param buf  for fmark->fnum.
+/// @param pos  for fmrak->mark.
+///
+/// @return[static] Mark with the given information.
+fmark_T *pos_to_mark(buf_T *buf, pos_T pos)
+{
+  static fmark_T mark = INIT_FMARK;
```
[the remark this answers is not in the document]
Comment (muniter) [reply src/nvim/mark.c]: Also wanted to point out that the current [implementation on master](https://github.com/neovim/neovim/blob/4b3fb4b17796ee6d9c8d7e8135a93b60cf09b016/src/nvim/mark.c#L337) is not save to cache pointers coming from mark. It also has a `static pos_T` that it returns sometimes.

File: src/nvim/mark_defs.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (jamessan) [reply src/nvim/mark_defs.h:15]: Ok.  Probably best to check the generated documentation to verify dev_style is correct. :)

File: src/nvim/mark.c [hunk 14]
```diff
@@ -221,245 +228,402 @@ void checkpcmark(void)
...
-pos_T *getmark_buf_fnum(buf_T *buf, int c, bool changefile, int *fnum)
+
+/// Get a local mark (lowercase and symbols).
+///
+/// Some marks are not actually marks, but positions that are never adjusted or motions presented as
+/// marks. It searches first for marks and falls back to finding motion type marks. If it's known
+/// ahead of time that the marks is actually a motion use the function mark_get_motion directly.
+///
+/// @note  Lowercase, last_cursor '"', last insert '^', last change '.' are not statically
+/// allocated.
```
[the remark this answers is not in the document]
Comment (jamessan) [reply src/nvim/mark.c:387]: > The freeing of memory related to marks is handled by the buffer creation, deletion logic.

The comment in the code is confusing to me because it makes me think of how I should be managing the memory that gets returned.  Your response makes it sound like the issue is more about whether the values can change out from under me, not about memory lifetimes.

If that's the case, I'd suggest rewording to focus more on this:

> The relevance is if you plan to get the value of multiple marks and you're not sure what type they are one should copy them (this is rare).

File: src/nvim/mark.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (jamessan) [reply src/nvim/mark.c]: > About caching the pointer I believe it shouldn't be done with marks, rather retrieve their value every time you need them. Reason being the buffer or window they are on can be gone any moment.

I wasn't thinking long-term caching, just having multiple calls in the same scope.  I'm not really familiar with how the mark APIs are used internally, so I was just commenting on what I see, even if the way the APIs are used makes them safe as is.

File: src/nvim/mark_defs.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (theHamsta) [reply src/nvim/mark_defs.h:15]: In the old days ///< was required for doxygen to refer to something before the comment. Maybe it has become more interligent

[the remark this answers is not in the document]
Comment (muniter) [reply src/nvim/mark_defs.h]: Well seems like it still needs it, red the docs and also checked https://neovim.io/doc/dev/index.html and saw that structs and enums without `<` were not documented. Just changed the comments.

File: src/nvim/mark.c [hunk 15]
```diff
@@ -221,245 +228,402 @@ void checkpcmark(void)
...
-pos_T *getmark_buf(buf_T *buf, int c, bool changefile)
+/// Get a named mark.
+///
+/// All types of marks, even those that are not technically a mark will be returned as such, to then
+/// move to the mark use mark_move_to().
+/// @param buf  Buffer to get the mark from.
+/// @param win  Window to get or calculate the mark from (motion type marks, context mark).
+/// @param flag  kMarkAll (default) returns any mark found, if global it resolves it's fnum.
+///              kMarkBufLocal return only marks that belong to buf, kMarkAllNoResolve same as
+///              kMarkAll but does not resolve the fnum.
```
[the remark this answers is not in the document]
Comment (muniter) [reply src/nvim/mark.c]: Went with the reference as suggested and improved the documentation there.

> Also, how is kMarkAll a default? Some value has to be passed in. Default implies there's a way to not specify any value.

Removed this, the enum documentation seems explanation enough.

File: src/nvim/mark.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (muniter) [reply src/nvim/mark.c:388]: Tweaked the documentation a bit, perhaps now is more clear.

**Comment by chentoast** [comment]
I hate to be “that person”, but are there plans to pick this up for 0.8? I’m asking because I have a bug fix that depends on the refactor done here.

**Comment by muniter** [comment]
> I hate to be “that person”, but are there plans to pick this up for 0.8? I’m asking because I have a bug fix that depends on the refactor done here.

Yes, been very busy but plan on picking this up very soon!

File: src/nvim/eval.c [hunk 16]
```diff
@@ -7807,12 +7807,15 @@ pos_T *var2fpos(const typval_T *const tv, const bool dollar_lnum, int *const ret
       pos = curwin->w_cursor;
     }
   } else if (name[0] == '\'') {
-    // mark
-    const pos_T *const pp = getmark_buf_fnum(curbuf, (uint8_t)name[1], false, ret_fnum);
-    if (pp == NULL || pp == (pos_T *)-1 || pp->lnum <= 0) {
+    // Mark.
```
Comment (zeertzjq) [inline src/nvim/eval.c:7810]:
```suggestion
    // mark
```

File: src/nvim/eval.c [hunk 17]
```diff
@@ -7807,12 +7807,15 @@ pos_T *var2fpos(const typval_T *const tv, const bool dollar_lnum, int *const ret
...
+    int mname = (uint8_t)name[1];
+    const fmark_T *const fm = mark_get(curbuf, curwin, NULL, kMarkAll, mname);
+    if (fm == NULL || fm->mark.lnum <= 0) {
       return NULL;
     }
-    pos = *pp;
+    pos = fm->mark;
+    // Vimscript behavior, only provide fnum if mark is global.
```
Comment (zeertzjq) [inline src/nvim/eval.c:7817]:
```suggestion
    // Vim script behavior, only provide fnum if mark is global.
```

**Comment by justinmk** [comment]
> 1. There's a new option. `jumpoptions+=view` enables the mark view restoring
>    automatically when using the jumplist and changelist. `^O`, `^I`, `g;`, `g,`

💯  will this also preserve the view with `<c-^>` ?

**Comment by muniter** [comment]
> 100 will this also preserve the view with `<c-^>` ?

@justinmk It didn't before, just implemented it in the last commit. Was thinking it could be a follow up but was very straightforward.

The work in this PR is done already, waiting for reviews, will try to keep it up to date.

I think there might be too many ways to restore the marks maybe we should pick only one `[number]'a` or `z'a` ?.

Mmm very weird things happening on CI, It's failing because of the test introduced in https://github.com/neovim/neovim/pull/18944/files but that commit is not on this branch history 😮

**Comment by zeertzjq** [comment]
checkout action checks out merge commit by default: <https://github.com/actions/checkout#checkout-pull-request-head-commit-instead-of-merge-commit=>

File: runtime/doc/options.txt [hunk 18]
```diff
@@ -3530,6 +3530,10 @@ A jump table for the options with a short description can be found at |Q_op|.
 			jumplist and then jumping to a location.
 			|jumplist-stack|
 
+	  view          When moving through the jumplist, changelist, and
+			|alternate-file| try to restore the |mark-view| in
+			which the jump occurred.
```
Comment (justinmk) [inline runtime/doc/options.txt:3535]: Why not also make it apply to `'x` ? Then we can drop the `z'` / `[count]'` variants, at least until we get strong feedback showing a use-case for _not_ restoring the view.

File: test/functional/editor/mark_spec.lua [hunk 19]
```diff
@@ -0,0 +1,432 @@
...
+    eq("Vim(normal):E20: Mark not set", err)
+  end)
+
+  it("errors when moving to a global mark that is not set with `", function()
+    command("edit " .. file1)
+    local err = pcall_err(helpers.exec_capture, "normal! `Z")
+    eq("Vim(normal):E20: Mark not set", err)
+  end)
+
+  it("can move to them using \'", function()
```
Comment (justinmk) [inline test/functional/editor/mark_spec.lua:98]:
```suggestion
  it("can move to them using '", function()
```

**Review by justinmk — APPROVED** [review]
Love this. Nice choices in naming/renaming, added comments, options, interface, verbosity, ...!

Let's wait a week for comments, then merge this.

Notes:

- is there a strong use case for the old behavior or was it an accident?
    - if the "restore view" behavior is controlled by 'jumpoptions' then do we still need `z'` / `[count]'` ?
- does the view also preserve the horizontal shift (or potentially in the future)?
- update `vim_diff.txt`

File: test/functional/ui/inccommand_spec.lua [hunk 20]
```diff
@@ -2970,7 +2970,7 @@ it(':substitute with inccommand, does nothing if range contains invalid marks',
     {15:~                             }|
     :'a,'bs/^                      |
   ]])
-  eq('', eval('v:errmsg'))
+  eq('E476: Invalid command', eval('v:errmsg'))
```
Comment (zeertzjq) [inline test/functional/ui/inccommand_spec.lua:2973]: ?

Comment (zeertzjq) [inline test/functional/ui/inccommand_spec.lua:2973]: ?

Comment (zeertzjq) [inline test/functional/ui/inccommand_spec.lua:2973]: Why did this change?

**Comment by muniter** [comment]
@justinmk 

> is there a strong use case for the old behavior or was it an accident?

I think the old behavior is just the way motions are implemented in the core editing model, from `motions.txt`:

> These commands move the cursor position.  If the new position is off of the
screen, **the screen is scrolled to show the cursor** (see also 'scrolljump' and
'scrolloff' options).

And the new "view" option enhances the "special" kind of motions marks are.

---

> if the "restore view" behavior is controlled by 'jumpoptions' then do we still need `z'` / `[count]'` ?

Went ahead and removed `z'` / `[count]'` and made 'jumpoptions' always set the view. At the beginning the strategy for restoring the view was different and I saw that some mappings of unimpaired.vim caused unnecessary scrolling, but now with the new strategy I don't see it anymore so more reason to remove.

> does the view also preserve the horizontal shift (or potentially in the future)?

Not currently, but seems simple in a followup. It's the main reason I went ahead with a struct in `fmarkv_T` even if using only one field right now.

> update `vim_diff.txt`

Done.

File: runtime/doc/motion.txt [hunk 21]
```diff
@@ -735,6 +735,11 @@ Jumping to a mark can be done in two ways:
 2. With ' (single quote): The cursor is positioned on the first non-blank
 			  character in the line of the specified location and
 			  the motion is linewise.
+						*mark-view*
+3. Apart from the above if 'jumpoptions' cotains "view", they will also try to
```
Comment (gpanders) [inline runtime/doc/motion.txt:739]:
```suggestion
3. Apart from the above if 'jumpoptions' contains "view", they will also try to
```

File: runtime/doc/vim_diff.txt [hunk 22]
```diff
@@ -383,6 +383,8 @@ Normal commands:
 Options:
   'ttimeout', 'ttimeoutlen' behavior was simplified
   |jumpoptions| "stack" behavior
+  |jumpoptions| "view" behavior which makes navigation less confusing, and
+  also applies to |mark-motions|.
```
Comment (gpanders) [inline runtime/doc/vim_diff.txt:387]: IMO "less confusing" is too vague and doesn't actually describe what `"view"` does.

Comment (gpanders) [reply runtime/doc/vim_diff.txt:387]: Agree about dropping `Which`, should also read `tries to restore` (missing `to`).

**Comment by muniter** [comment]
@justinmk friendly ping. :)

**Comment by justinmk** [comment]
coverity issue:

```
*** CID 353827:  Null pointer dereferences  (REVERSE_INULL)
/src/nvim/mark.c: 283 in get_jumplist()
277             count += count < 0 ? -1 : 1;
278             continue;
279           }
280         }
281         break;
282       }
>>>     CID 353827:  Null pointer dereferences  (REVERSE_INULL)
>>>     Null-checking "jmp" suggests that it may be null, but it has already been dereferenced on all paths leading to the check.
283       return jmp != NULL ? &jmp->fmark : NULL;
284     }
285     
286     /// Get mark in "count" position in the |changelist| relative to the current index.
287     ///
288     /// @note  Changes the win->w_changelistidx.
```
